app/actions/client.py

```diff
@@ -0,0 +1,102 @@
+import logging
+
+from datetime import datetime
+
+import pydantic
+import httpx
+from pydantic import root_validator
+from typing import List, Optional
+
+from app.services.state import IntegrationStateManager
+
+
+state_manager = IntegrationStateManager()
+logger = logging.getLogger(__name__)
+
+
+# Exception classes
+class DigitAnimalUnauthorizedException(Exception):
+    def __init__(self, message: str, error: Exception = None, status_code=401):
+        self.status_code = status_code
+        self.message = message
+        self.error = error
+        super().__init__(f"'{self.status_code}: {self.message}, Error: {self.error}'")
+
+
+class DigitAnimalErrorException(Exception):
+    def __init__(self, message: str, error: Exception = None, status_code=500):
+        self.status_code = status_code
+        self.message = message
+        self.error = error
+        super().__init__(f"'{self.status_code}: {self.message}, Error: {self.error}'")
+
+
+# Pydantic models (representing integration objects to receive/manipulate info from the external API)
+class DigitAnimalHistoricalRequestParams(pydantic.BaseModel):
+    init_date: str
+    end_date: str
+
+    @root_validator
+    def parse_datetime(cls, values):
+        for key, val in values.items():
+            values[key] = val.strftime("%Y-%m-%d %H:%M:%S")
+        return values
+
+
+class DigitAnimalDataResponse(pydantic.BaseModel):
+    DEVICE_COLLAR: str
+    LAT: float
+    LNG: float
+    DEVICE_TIME: datetime
+    DEVICE_ALARM: Optional[bool]
+    DEVICE_LOCATION: Optional[bool]
+    DEVICE_TEMPERATURE: Optional[bool]
+    DEVICE_DISTANCE: Optional[bool]
+    DEVICE_ACTIVITY: Optional[bool]
+    DEVICE_POSITION: Optional[bool]
+    RAW_TEMPERATURE: Optional[float]
+    RAW_ACC_X: Optional[float]
+    RAW_ACC_Y: Optional[float]
+    RAW_ACC_Z: Optional[float]
+
+
+class DigitAnimalData(pydantic.BaseModel):
+    devices: Optional[List[DigitAnimalDataResponse]]
+    history: List[DigitAnimalDataResponse]
+
+
+class DigitAnimalResponse(pydantic.BaseModel):
+    success: bool
+    message: str
+    data: DigitAnimalData
+
+
+async def get_devices_observations(integration_id, base_url, auth, params=None):
```

Comment:
I recommend adding type hints in general and especially for the clients. That'll help us avoid some mysterious errors as these things age and have modifications.